subsys/partition_manager/Kconfig

```diff
@@ -174,6 +174,15 @@ config PM_PARTITION_SIZE_TFM_EXTRA
 
 endif
 
+config TFM_BOARD
+	string
+	default "../../../../../../../nrf/subsys/tfm/boards/nrf9160" if SOC_NRF9160_SICA
```

Comment:
we should not use relative paths in Kconfig, cause it is never clear what it is relative to.
Sometimes people think relative to a conf file, sometimes to app dir, and in this case I assume you consider relative to a folder inside the tfm repo going into nrf, making thing break if we relocate the TFM module later (as example going from 3 -> single repo clone or switching between nrf/zephyr fork and vanilla upstream).

But relative is very unclear when you start placing settings in Kconfig fragments, or specify them using `-DCONFIG_<foo>`.

Why does this value need to be a path in Kconfig ?
Could we not keep the upstream symbol, and then in CMake, adjust the value of `CONFIG_TFM_BOARD` like:
```
set(CONFIG_TFM_BOARD ${OUR_PREFIX_WITH_PATH}/${CONFIG_TFM_BOARD})
```
would of course require that we put out boards in `nordic_nrf/nrf9160dk_nrf9160` instead of just `nrf9160`.

Reply:
As example, you may do:
```
file(RELATIVE_PATH
     TFM_NRF_RELATIVE_TARGET_DIR
     ${ZEPHYR_TFM_MODULE_DIR}/trusted-firmware-m/platform/ext/target/
     ${NRF_DIR}/subsys/tfm/boards
)
set(CONFIG_TFM_BOARD ${TFM_NRF_RELATIVE_TARGET_DIR}/${CONFIG_TFM_BOARD})
```

which will set: `TFM_NRF_RELATIVE_TARGET_DIR` to `../../../../../../../nrf/subsys/tfm/boards`

Reply:
@tejlmand I created a new Kconfig TFM_BOARD_BASE_DIR. Working with variables in Cmake gets tricky because you don't know quite which order things will happen in. I opted for KConfig string that can contain Cmake variables. This means that the Cmake context is known (TFM module), but the value can still be overridden from anywhere. Tell me what you think

Reply:
>  @tejlmand I created a new Kconfig TFM_BOARD_BASE_DIR.

Thanks for the proposal. I see it requires an additional patch upstream, and in that case, I think it would be even better if TFM itself could be improved to allow out of tree boards.
This setting is similar to a board root in Zephyr, which is something we normally don't put in Kconfig, but of course it is a little different here, as this is a Zephyr module.

> Working with variables in Cmake gets tricky because you don't know quite which order things will happen in. 
> I opted for KConfig string that can contain Cmake variables. 

Why is this less tricky ?
Now you are  depending on Kconfig processing order.

You are depending upon the fact that this line:
https://github.com/nrfconnect/sdk-nrf/blob/568fb2fa45e129927f5d148877e788ef285ce1fa/subsys/tfm/Kconfig#L17-L19

must be processed before this line:
https://github.com/nrfconnect/sdk-zephyr/blob/912208a6d5fef1e01bc1813efed76b615140e1ac/modules/Kconfig.tfm#L23
```
config TFM_BOARD_BASE_DIR
	string
	default "${TFM_BOARD_BASE_DIR_DEFAULT}" 
```

and do you know how the process order Zephyr module Kconfig files are ensured compared to Zephyr module CMakeLists.txt files ?
They are processed in exactly same order, so that if `moduleA` is processed before `moduleB` in Kconfig, then the same is true for the `moduleA` and `moduleB` in CMake. (Of course for Zephyr modules having both a CMakeLists.txt and Kconfig file)

The Kconfig process order of modules is placed in `<build>/Kconfig/Kconfig.modules`:
```
osource "/projects/github/ncs/nrf/Kconfig.nrf"
osource "/projects/github/ncs/bootloader/mcuboot/zephyr/Kconfig"
osource "/projects/github/ncs/nrfxlib/Kconfig.nrfxlib"
```
and thereafter all Kconfig files in `$ZEPHYR_BASE/modules/Kconfig.<module>`.

and CMake in `<build>/zephyr_modules.txt`:
```
"nrf":"/projects/github/ncs/nrf":"/projects/github/ncs/nrf"
"mcuboot":"/projects/github/ncs/bootloader/mcuboot":"/projects/github/ncs/bootloader/mcuboot/zephyr"
...
"nrfxlib":"/projects/github/ncs/nrfxlib":"/projects/github/ncs/nrfxlib"
"tfm":"/projects/github/ncs/modules/tee/tfm":"/projects/github/ncs/modules/tee/tfm"
...
```

and this order is same order as `west list` adjusted for any `depends` dependencies, so that those dependencies are guaranteed.

This means the manifest project is first in list, then local defined projects, then imported projects.
Unless a module is using `depends`, which means it will of course be processed after the module on which it depends.
https://docs.zephyrproject.org/latest/guides/modules.html#zephyr-module-dependencies

So if this is your only concern, I would stick to the CMake, and avoid the additional patches.

Note: the fact that all Kconfig files in `$ZEPHYR_BASE/modules/Kconfig.<module>` are processed in a bunh, is because they are not truly mdule code, which is one of the reasons for fixing this as part of: https://github.com/zephyrproject-rtos/zephyr/pull/30904

Reply:
I get your point about resolving order. But is modifying a CONFIG_* variable in Cmake really the best way to solve this? That strikes me as hacky :P.

Reply:
> > That strikes me as hacky :P.

I agree on this, and as I mentioned, I believe  that we should aim at pushing a proper solution to TFM, as I wrote earlier:
> I think it would be even better if TFM itself could be improved to allow out of tree boards.

then the corresponding in Zephyr would not need to be a path, but could be a choice or similar config that would translate into a TFM board dir / root to be passed on to TFM build system.
Such a setting translation could then be defined by the module itself, in this case nrf.

With such a goal, I think a pure NCS local fix / hack of proposed nature would introduce the minimal amount of changes locally, and require no upstream patches. Patches that would anyway have to be re-worked when TFM truly supports out-of-tree boards.

Reply:
Ok, I'm working on a proposal for upstream where it's possible to pass a platform as an absolute path. Do you see any problems with that?

Reply:
> Do you see any problems with that?

Guess I would need to see the proposal, in order to tell :wink: 
But it sounds like a good idea.

Reply:
Ok, cool. Take a look: https://review.trustedfirmware.org/c/TF-M/trusted-firmware-m/+/8064

Reply:
This is now merged and brought into https://github.com/nrfconnect/sdk-trusted-firmware-m/pull/1